perf-tool/src/server.cpp

```diff
@@ -109,133 +143,178 @@ void Server::handleServer()
+                            pollFds[socketIndex].fd = -1; /* TODO: we need a linked list rather than an array */
+                        }
+                    }
+                    else if (pollFds[ServerConstants::BASE_POLLS + activeNetworkClients].revents & POLLIN)
+                    {
+                        command = networkClients[i]->handlePoll();
+                        if (!command.empty())
+                        {
+                            handleClientCommand(command, "Client");
+                        }
+                    }
+                }
+                /* Now, handle new connection requests */
+                if (pollFds[0].revents & POLLIN)
+                {
+                    /* The accept() call actually accepts an incoming connection */
+                    clilen = sizeof(cli_addr);
+
+                    /* This accept() function will write the connecting client's address info 
+                     * into the the address structure and the size of that structure is clilen.
+                     */
+                    newsocketFd = accept(serverSocketFd, (struct sockaddr *)&cli_addr, &clilen);
+                    if (newsocketFd < 0)
+                    {
+                        perror("ERROR on accept");
+                    }
+                    else
+                    {
+                        /* Check the upper limit on active connections */
+                        if (activeNetworkClients < ServerConstants::NUM_CLIENTS)
+                        {
+                            networkClients[activeNetworkClients] = new NetworkClient(newsocketFd);
+                            if (verbose >= Verbose::INFO)
+                                printf("server: got connection from %s port %d\n",
+                                        inet_ntoa(cli_addr.sin_addr),
+                                        ntohs(cli_addr.sin_port));
+
+                            /* Send a welcome message */
+                            sendMessage(newsocketFd, "Connection to server succeeded", "ServerStartEvent", "Server");
+
+                            /* Update number of active clients */
+                            activeNetworkClients++;
+                        }
+                        else /* Too many connections */
+                        {
+                            close(newsocketFd);
+                            if (verbose >= Verbose::WARN)
+                                printf("server refusing connection because there are too many\n");
+                        }
+                    }
+                } 
             }
         }
-        /* Check for commands from commands file if running in headless mode */
-        if (headlessMode)
+    }
+    else /* no network commands are allowed */
+    {
+        if (headlessMode && !delayedCommands.empty())
         {
-            jsonCommand = commandClient->handlePoll();
-            if (!jsonCommand.empty())
+            while(1) 
```

Comment:
From the looks of things, `handleServer` is called essentially in the VM_INIT hook or in Agent_OnAttach, which means the init thread will block until we're done all the delayed commands. Is this intentional, or is the delayed commands meant to happen in parallel with JVM init?

Reply:
VM_INIT starts a new thread for the server and `handleServer` is executed by this new thread:
```
JNIEXPORT void JNICALL VMInit(jvmtiEnv *jvmtiEnv, JNIEnv* jni_env, jthread thread) 
{
    jvmtiError error;
    server = new Server(portNo, commandsPath, logPath);

    error = jvmtiEnv -> RunAgentThread( createNewThread(jni_env), &startServer, NULL, JVMTI_THREAD_NORM_PRIORITY );
    check_jvmti_error_throw(jvmtiEnv, error, "Error starting agent thread.");
    if (verbose >= WARN)
        printf("VM starting up.\n");
}
```